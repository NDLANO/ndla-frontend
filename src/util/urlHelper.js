/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { matchPath } from 'react-router-dom';
import parseUrl from 'parse-url';
import { isValidLocale } from '../i18n';
import {
  RESOURCE_PAGE_PATH,
  PLAIN_ARTICLE_PAGE_PATH,
  TOPIC_PATH,
} from '../constants';

function matchUrl(pathname, type, lang = false) {
  if (type === 'article') {
    return matchPath(
      pathname,
      lang ? `/:lang${PLAIN_ARTICLE_PAGE_PATH}` : PLAIN_ARTICLE_PAGE_PATH,
    );
  }
  let match = matchPath(
    pathname,
    lang ? `/:lang${RESOURCE_PAGE_PATH}` : RESOURCE_PAGE_PATH,
  );
  console.log('nfhfhr', match);
  if (match) {
    return match;
  }

  return matchPath(pathname, lang ? `/:lang${TOPIC_PATH}` : TOPIC_PATH);
}

export function parseAndMatchUrl(url) {
  const { pathname } = parseUrl(url);
  const paths = pathname.split('/');
  paths[1] = paths[1] === 'unknown' ? 'nb' : paths[1];
  const path = paths.join('/');

  if (isValidLocale(paths[1])) {
    return matchUrl(path, paths[2], true);
  }
  return matchUrl(path, paths[1], false);
}
