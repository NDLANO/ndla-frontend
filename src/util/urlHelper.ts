/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { matchPath, Params, PathMatch, Location } from "react-router-dom";
import { GQLTaxBase } from "../graphqlTypes";
import { isValidLocale, supportedLanguages } from "../i18n";
import { oembedRoutes } from "../routes";

type OembedParams =
  | "subjectId"
  | "topicId"
  | "resourceId"
  | "stepId"
  | "contextId"
  | "articleId"
  | "lang"
  | "topicOrResourceId"
  | "audioId"
  | "videoId"
  | "imageId"
  | "conceptId"
  | "h5pId";

type OembedReturnParams =
  | "subjectId"
  | "topicId"
  | "resourceId"
  | "stepId"
  | "contextId"
  | "articleId"
  | "lang"
  | "audioId"
  | "videoId"
  | "imageId"
  | "conceptId"
  | "h5pId";

const matchUrl = (pathname: string, lang: boolean = false): PathMatch<OembedReturnParams> | null => {
  const possiblePaths = lang ? oembedRoutes.map((r) => `/:lang/${r}`) : oembedRoutes;

  let match: PathMatch<OembedParams> | undefined;
  for (let i = 0; i < possiblePaths.length; i++) {
    const attempt = possiblePaths[i] ?? "";
    const m = matchPath<OembedParams, string>(attempt, pathname);
    if (m) {
      match = m;
      break;
    }
  }
  if (!match) {
    return null;
  }

  if (match.params.topicOrResourceId) {
    const missingParam: {
      resourceId?: string;
      topicId?: string;
    } = match.params.topicOrResourceId.startsWith(":resource")
      ? { resourceId: match.params.topicOrResourceId.replace(":resource", "") }
      : { topicId: match.params.topicOrResourceId.replace(":topic", "") };
    const params: Params<OembedParams> = match.params;
    return {
      ...match,
      params: {
        ...params,
        ...missingParam,
      },
    };
  } else return match;
};

export function parseOembedUrl(url: string, ignoreLocale: boolean = false) {
  const { pathname } = new URL(url);
  const paths = pathname.split("/");
  if (paths[1]) {
    paths[1] = paths[1] === "unknown" ? "nb" : paths[1];
  }
  if (ignoreLocale && isValidLocale(paths[1])) {
    paths.splice(1, 1);
  }
  const path = paths.join("/");

  if (isValidLocale(paths[1])) {
    return matchUrl(path, true);
  }
  return matchUrl(path, false);
}

export const toHref = (location: Location) => {
  return `${location.pathname}${location.search}`;
};

export const constructNewPath = (pathname: string, newLocale?: string) => {
  const regex = new RegExp(`\\/(${supportedLanguages.join("|")})($|\\/)`, "");
  const path = pathname.replace(regex, "");
  const fullPath = path.startsWith("/") ? path : `/${path}`;
  const localePrefix = newLocale ? `/${newLocale}` : "";
  return `${localePrefix}${fullPath}`;
};

export const isCurrentPage = (pathname: string, taxBase: Pick<GQLTaxBase, "path" | "url">) => {
  let path = pathname.replace(/\/$/, ""); // Remove trailing slash if present
  const match = matchUrl(path);
  if (match?.params.stepId) {
    path = path.replace(/\/\d+$/, ""); // Remove last numeric segment if stepId
  }
  return path === taxBase.path || decodeURIComponent(path) === taxBase.url;
};
