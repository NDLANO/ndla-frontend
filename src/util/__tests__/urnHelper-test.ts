/**
 * Copyright (c) 2019-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { parseOembedUrl } from "../urlHelper";

const validArticleUrl1 = "https://www.test.ndla.no/subject:3/topic:1:55163/topic:1:168398/resource:1:1682";
const validArticleUrl2 = "http://localhost:3000/subject:3/topic:1:168542/topic:1:173292/resource:1:168554";
const validArticleUrl3 = "http://localhost:3000/subject:3/topic:1:168542/resource:1:168554";
const validArticleUrl4 = "https://www.test.ndla.no/unknown/subject:3/topic:1:55163/topic:1:168398/resource:1:1682";

const validArticleUrlWithLang = "https://www.test.ndla.no/nn/subject:3/topic:1:55163/topic:1:168398/resource:1:1682";
const validTopicArticleUrl = "https://www.test.ndla.no/subject:3/topic:1:55163";
const unvalidArticleUrl = "https://www.test.ndla.no/subject:3";

const validateSimpleArticlePath1 = "https://www.test.ndla.no/article/4809";
const validateSimpleArticlePath2 = "https://www.test.ndla.no/nn/article/4809";

const unvalidateSimpleArticlePath = "https://www.test.ndla.no/articles/4809";

test("parseAndMatchUrl", () => {
  expect(parseOembedUrl(validArticleUrl1)).toMatchSnapshot();
  expect(parseOembedUrl(validTopicArticleUrl)).toMatchSnapshot();
  expect(parseOembedUrl(validArticleUrl2)).toMatchSnapshot();
  expect(parseOembedUrl(validArticleUrl3)).toMatchSnapshot();
  expect(parseOembedUrl(validArticleUrl4)).toMatchSnapshot();
  expect(parseOembedUrl(validArticleUrlWithLang)).toMatchSnapshot();
  expect(parseOembedUrl(unvalidArticleUrl)).toBe(null);
  expect(parseOembedUrl(validateSimpleArticlePath1)).toMatchSnapshot();
  expect(parseOembedUrl(validateSimpleArticlePath2)).toMatchSnapshot();
  expect(parseOembedUrl(unvalidateSimpleArticlePath)).toBe(null);
});
