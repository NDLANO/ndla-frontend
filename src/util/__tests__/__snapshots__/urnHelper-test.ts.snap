// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseAndMatchUrl 1`] = `
Object {
  "params": Object {
    "resourceId": ":1:1682",
    "subjectId": ":3",
    "topic1": ":1:55163",
    "topicId": ":1:168398",
  },
  "pathname": "/subject:3/topic:1:55163/topic:1:168398/resource:1:1682",
  "pathnameBase": "/subject:3/topic:1:55163/topic:1:168398/resource:1:1682",
  "pattern": Object {
    "caseSensitive": false,
    "end": true,
    "path": "subject:subjectId/topic:topic1/topic:topicId/resource:resourceId",
  },
}
`;

exports[`parseAndMatchUrl 2`] = `
Object {
  "params": Object {
    "subjectId": ":3",
    "topicId": ":1:55163",
  },
  "pathname": "/subject:3/topic:1:55163",
  "pathnameBase": "/subject:3/topic:1:55163",
  "pattern": Object {
    "caseSensitive": false,
    "end": true,
    "path": "subject:subjectId/topic:topicId",
  },
}
`;

exports[`parseAndMatchUrl 3`] = `
Object {
  "params": Object {
    "resourceId": ":1:168554",
    "subjectId": ":3",
    "topic1": ":1:168542",
    "topicId": ":1:173292",
  },
  "pathname": "/subject:3/topic:1:168542/topic:1:173292/resource:1:168554",
  "pathnameBase": "/subject:3/topic:1:168542/topic:1:173292/resource:1:168554",
  "pattern": Object {
    "caseSensitive": false,
    "end": true,
    "path": "subject:subjectId/topic:topic1/topic:topicId/resource:resourceId",
  },
}
`;

exports[`parseAndMatchUrl 4`] = `
Object {
  "params": Object {
    "resourceId": ":1:168554",
    "subjectId": ":3",
    "topicId": ":1:168542",
  },
  "pathname": "/subject:3/topic:1:168542/resource:1:168554",
  "pathnameBase": "/subject:3/topic:1:168542/resource:1:168554",
  "pattern": Object {
    "caseSensitive": false,
    "end": true,
    "path": "subject:subjectId/topic:topicId/resource:resourceId",
  },
}
`;

exports[`parseAndMatchUrl 5`] = `
Object {
  "params": Object {
    "lang": "nb",
    "resourceId": ":1:1682",
    "subjectId": ":3",
    "topic1": ":1:55163",
    "topicId": ":1:168398",
  },
  "pathname": "/nb/subject:3/topic:1:55163/topic:1:168398/resource:1:1682",
  "pathnameBase": "/nb/subject:3/topic:1:55163/topic:1:168398/resource:1:1682",
  "pattern": Object {
    "caseSensitive": false,
    "end": true,
    "path": "/:lang/subject:subjectId/topic:topic1/topic:topicId/resource:resourceId",
  },
}
`;

exports[`parseAndMatchUrl 6`] = `
Object {
  "params": Object {
    "lang": "nn",
    "resourceId": ":1:1682",
    "subjectId": ":3",
    "topic1": ":1:55163",
    "topicId": ":1:168398",
  },
  "pathname": "/nn/subject:3/topic:1:55163/topic:1:168398/resource:1:1682",
  "pathnameBase": "/nn/subject:3/topic:1:55163/topic:1:168398/resource:1:1682",
  "pattern": Object {
    "caseSensitive": false,
    "end": true,
    "path": "/:lang/subject:subjectId/topic:topic1/topic:topicId/resource:resourceId",
  },
}
`;

exports[`parseAndMatchUrl 7`] = `
Object {
  "params": Object {
    "articleId": "4809",
  },
  "pathname": "/article/4809",
  "pathnameBase": "/article/4809",
  "pattern": Object {
    "caseSensitive": false,
    "end": true,
    "path": "article/:articleId",
  },
}
`;

exports[`parseAndMatchUrl 8`] = `
Object {
  "params": Object {
    "articleId": "4809",
    "lang": "nn",
  },
  "pathname": "/nn/article/4809",
  "pathnameBase": "/nn/article/4809",
  "pattern": Object {
    "caseSensitive": false,
    "end": true,
    "path": "/:lang/article/:articleId",
  },
}
`;
